examples/bovw/plot_bovw.py

```diff
@@ -0,0 +1,154 @@
+performance.
+
+References
+----------
+.. [1] Varma, Manik, and Andrew Zisserman. "A statistical approach to material
+       classification using image patch exemplars." IEEE transactions on
+       pattern analysis and machine intelligence 31.11 (2009): 2032-2047.
+
+"""
+
+# Author: Guillaume Lemaitre <g.lemaitre58@gmail.com>
+# License: BSD
+
+import numpy as np
+from scipy.misc import imread
+
+from sklearn.feature_extraction.image import extract_patches_2d
+from sklearn.decomposition import PCA
+from sklearn.model_selection import StratifiedKFold
+from sklearn.cluster import MiniBatchKMeans
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.metrics import confusion_matrix
+from sklearn.externals.joblib import Parallel, delayed
+
+from tudarmstadt import fetch_tu_darmstadt
+
+
+print(__doc__)
+
+
+def proc_image(path_image, rng, patch_size=(9, 9), max_patches=10000):
+    """ Function to extract couple of patches from an image """
+
+    # Read the current image
+    im = imread(path_image)
+    # Extract patches
+    patch = extract_patches_2d(im, patch_size=patch_size,
+                               max_patches=max_patches, random_state=rng)
+    return patch.reshape((max_patches, np.prod(patch_size) * len(im.shape)))
+
+
+def image_extraction_projection(path_image, dict_PCA, rng,
+                                patch_size=(9, 9), max_patches=100000):
+    """ Function to extract a couple of patches from an image and apply PCA """
+
+    # Read the current image
+    im = imread(path_image)
+    # Extract patches
+    patch = extract_patches_2d(im, patch_size=patch_size,
+                               max_patches=max_patches, random_state=rng)
+
+    return dict_PCA.transform(patch.reshape((patch.shape[0],
+                                             np.prod(patch_size) *
+                                             len(im.shape))))
+
+# Define the parameters in use afterwards
+patch_size = (9, 9)
+max_patches = 100
+n_jobs = -1
+n_components = 9
+max_patches_classify = 2000
+nb_words = 50
+rng = 42
+
+# Load the data
+png_files, labels = fetch_tu_darmstadt()
+
+# We will extract some patches from the different images to construct
+# a PCA codebook that will be used afterwards
+# Extract the patches
+patch_arr = Parallel(n_jobs=n_jobs)(delayed(proc_image)(path_im,
+                                                        rng,
+                                                        patch_size,
+                                                        max_patches)
+                                    for path_im in png_files)
+
+print('Extracted patch to build dictionary')
+
+# Create a plain matrix to apply the PCA decomposition
+patch_arr = np.array(patch_arr, copy=False)
+patch_arr = patch_arr.reshape((patch_arr.shape[0] * patch_arr.shape[1],
+                               patch_arr.shape[2]))
+# Build a PCA dictionary
+dict_PCA = PCA(n_components=n_components, random_state=rng)
```

Comment:
The results are equivalent. I remove the PCA since it shorten the example and this is still a texton approach.